src/EditorFeatures/CSharp/DecompiledSource/AssemblyResolver.cs

```diff
@@ -109,8 +111,33 @@ PEFile MakePEFile(IAssemblySymbol assembly)
                 // reference assemblies should be fine here, we only need the metadata of references.
                 var reference = _parentCompilation.GetMetadataReference(assembly);
                 Log(CSharpEditorResources.Load_from_0, reference.Display);
-                return new PEFile(reference.Display, PEStreamOptions.PrefetchMetadata);
+                if (File.Exists(reference.Display))
+                {
+                    return new PEFile(reference.Display, PEStreamOptions.PrefetchMetadata);
+                }
+                else
+                {
+                    var peReader = GetPropertyValue<PEReader>(GetPropertyValue(assembly.GetMetadata().GetModules().First(), "Module"), "PEReaderOpt");
```

Comment:
@dotnet/roslyn-compiler Can you point me to the non-reflection way to achieve this?

Reply:
> Can you point me to the non-reflection way to achieve this?

In the short time I spend on this I didn't find a public API to get this information. More over, I believe this is done on purpose. We probably don't want to expose compiler's internal implementation details like this. There also could be some lifetime issues with exposing these internal objects. Bottom line, what you are doing feels dangerous to me. @tmat Do you have any comments?

---
In reply to: [605371535](https://github.com/dotnet/roslyn/pull/52318#discussion_r605371535) [](ancestors = 605371535)

Reply:
You need to cast reference to PortableExecutableReference. Then you can use FilePath or GetMetadata(). 
A reference does not need to be PE Reference, it could be a compilation reference. 
FilePath may not be available.
The PE Image might not be available - we might only have the metadata section, so only MetadataReader may be available.

---
In reply to: [605974479](https://github.com/dotnet/roslyn/pull/52318#discussion_r605974479) [](ancestors = 605974479,605371535)

Reply:
@tmat for the case I need, the PortableExecutableReference points to an image in memory, and there is no corresponding file on disk. I need the `PEReader` for this image so the feature can work.

Reply:
That's gonna be a problem since we don't generally load IL in memory, only metadata. How came we don't have a file path for a metadata reference? We ought to have it in the project.

Reply:
The assembly was emitted from a Compilation directly to memory, and never saved to disk.

Reply:
Where was it emitted?

---
In reply to: [606329820](https://github.com/dotnet/roslyn/pull/52318#discussion_r606329820) [](ancestors = 606329820)

Reply:
https://github.com/dotnet/roslyn/blob/6546dbc44406b1dee5238f33b26264fc8ce2146b/src/Compilers/Test/Core/Compilation/CompilationExtensions.cs#L44

Reply:
If this is for testing only then let's update the tests to write the file on disk.

Reply:
➡️ This is now rewritten